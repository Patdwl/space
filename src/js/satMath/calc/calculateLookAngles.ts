import { DEG2RAD } from '@app/js/lib/constants';
import { calcSatrec } from '@app/js/satSet/catalogSupport/calcSatrec';
import { keepTrackApi } from '../../api/keepTrackApi';
import { SatObject, SensorObject, TearrData } from '../../api/keepTrackTypes';
import { satellite } from '../satMath';
import { getTearData } from './getTearData';

export const calculateLookAngles = (sat: SatObject, sensors: SensorObject[]): TearrData[] => {
  const { sensorManager, timeManager } = keepTrackApi.programs;

  (function _inputValidation() {
    // Check if there is a sensor
    if (typeof sensors == 'undefined') {
      // Try using the current sensor if there is one
      if (sensorManager.checkSensorSelected()) {
        sensors = sensorManager.currentSensor;
      } else {
        console.debug('getlookangles2 requires a sensor!');
        return;
      }
      // Simple Error Checking
    } else {
      if (typeof sensors[0].obsminaz == 'undefined') {
        console.debug('sensors[0] format incorrect');
        return;
      }
      sensors[0].observerGd = {
        // Array to calculate look angles in propagate()
        lat: sensors[0].lat * DEG2RAD,
        lon: sensors[0].lon * DEG2RAD,
        alt: sensors[0].alt,
      };
    }

    if (typeof sat == 'undefined') {
      console.debug('sat parameter required!');
    } else {
      if (typeof sat.TLE1 == 'undefined' || typeof sat.TLE2 == 'undefined') {
        console.debug('sat parameter invalid format!');
      }
    }

    if (typeof satellite.isRiseSetLookangles == 'undefined') {
      satellite.isRiseSetLookangles = false;
    }
  })();

  // TOOD: Instead of doing the first sensor this should return an array of TEARRs for all sensors.
  const sensor = sensors[0];

  const simulationTime = timeManager.simulationTimeObj;
  let offset = 0;
  var satrec = calcSatrec(sat);
  var lookanglesTable = []; // Iniially no rows to the table
  var tempLookanglesInterval;

  if (satellite.isRiseSetLookangles) {
    tempLookanglesInterval = satellite.lookanglesInterval;
    satellite.lookanglesInterval = 1;
  }

  for (var i = 0; i < satellite.lookanglesLength * 24 * 60 * 60; i += satellite.lookanglesInterval) {
    // satellite.lookanglesInterval in seconds
    offset = i * 1000; // Offset in seconds (msec * 1000)
    const now = timeManager.getOffsetTimeObj(offset, simulationTime);
    if (lookanglesTable.length <= 5000) {
      // Maximum of 1500 lines in the look angles table
      const _lookanglesRow = getTearData(now, satrec, [sensor], satellite.isRiseSetLookangles);
      if (_lookanglesRow.time !== '') {
        lookanglesTable.push(_lookanglesRow); // Update the table with looks for this 5 second chunk and then increase table counter by 1
      }
    }
  }

  if (satellite.isRiseSetLookangles) {
    satellite.lookanglesInterval = tempLookanglesInterval;
  }
  return lookanglesTable;
};
