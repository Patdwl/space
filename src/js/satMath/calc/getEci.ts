import { calcSatrec } from '@app/js/satSet/catalogSupport/calcSatrec';
import { Sgp4 } from 'ootk';
import { SatObject } from '../../api/keepTrackTypes';
import { calculateTimeVariables } from './calculateTimeVariables';

export const getEci = (sat: SatObject, now: Date) => {
  try {
    let satrec = calcSatrec(sat);
    const { m } = calculateTimeVariables(now, satrec);

    return Sgp4.propagate(satrec, m);
  } catch {
    return { position: { x: 0, y: 0, z: 0 }, velocity: { x: 0, y: 0, z: 0 } };
  }
};
