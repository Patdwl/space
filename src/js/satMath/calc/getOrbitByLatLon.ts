import { SatObject } from '@app/js/api/keepTrackTypes';
import { RAD2DEG } from '@app/js/lib/constants';
import { stringPad } from '@app/js/lib/helpers';
import { EciVec3, SatelliteRecord } from 'ootk';
import { satellite } from '../satMath';
import { calculateTimeVariables } from './calculateTimeVariables';

enum PropagationOptions {
  MeanAnomaly = 1,
  RightAscensionOfAscendingNode = 2,
  ArgumentOfPerigee = 3,
}

enum PropagationResults {
  Near = 0,
  Success = 1,
  Error = 2,
  Far = 3,
}

// prettier-ignore
export const getOrbitByLatLon = (
  sat: SatObject,
  goalLat: number,
  goalLon: number,
  goalDirection: 'N' | 'S',
  now: Date,
  goalAlt?: number,
  raanOffset?: number
): [string, string] => { // NOSONAR
  let newMeana: string = null;
  let newArgPer: string = null;
  goalAlt = goalAlt || null;
  raanOffset = raanOffset || 0;

  let lastLat: number = null;
  let currentDirection: 'N' | 'S';

  // Parse some values used in creating new TLEs
  const intl = sat.TLE1.substr(9, 8);
  const epochyr = sat.TLE1.substr(18, 2);
  const epochday = sat.TLE1.substr(20, 12);
  const meanmo = sat.TLE2.substr(52, 11);
  const inc = stringPad.pad0((sat.inclination * RAD2DEG).toFixed(4), 8);
  const ecen = sat.eccentricity.toFixed(7).substr(2, 7);
  // Disregarding the first and second derivatives of mean motion
  // Just keep whatever was in the original TLE
  const TLE1Ending = sat.TLE1.substr(32, 39);

  // //////////////////////////////////////////////////////////////////////////////
  // //////////////////////////////////////////////////////////////////////////////
  // These four functions get called during the bottom loop

  /**
   * Rotating the mean anomaly adjusts the latitude (and longitude) of the satellite.
   * @param {number} meana - This is the mean anomaly (where it is along the orbital plane)
   * @returns {PropagationResults} This number tells the main loop what to do next
   */
  const meanaCalc = (meana: number) => {
    let satrec = sat.satrec || <SatelliteRecord>satellite.twoline2satrec(sat.TLE1, sat.TLE2); // perform and store sat init calcs

    meana = meana / 10;
    const meanaStr = stringPad.pad0(meana.toFixed(4), 8);

    const raan = stringPad.pad0((sat.raan * RAD2DEG).toFixed(4), 8);

    const argPe = newArgPer ? stringPad.pad0((parseFloat(newArgPer) / 10).toFixed(4), 8) : stringPad.pad0((sat.argPe * RAD2DEG).toFixed(4), 8);

    const _TLE1Ending = sat.TLE1.substr(32, 39);
    const TLE1 = '1 ' + sat.sccNum + 'U ' + intl + ' ' + epochyr + epochday + _TLE1Ending; // M' and M'' are both set to 0 to put the object in a perfect stable orbit
    const TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + raan + ' ' + ecen + ' ' + argPe + ' ' + meanaStr + ' ' + meanmo + '    10';

    satrec = satellite.twoline2satrec(TLE1, TLE2);
    const results = getOrbitByLatLonPropagate(now, satrec, PropagationOptions.MeanAnomaly);
    if (results === PropagationResults.Success) {
      sat.TLE1 = TLE1;
      sat.TLE2 = TLE2;
      newMeana = meanaStr;
    }
    return results;
  };

  /**
   * We need to adjust the argument of perigee to align a HEO orbit with the desired launch location
   * @param {string} argPe - This is the guess for the argument of perigee (where the lowest part of the orbital plane is)
   * @returns {PropagationResults} This number tells the main loop what to do next
   */
  const argPerCalc = (argPe: string): PropagationResults => {
    const meana = newMeana;
    const raan = stringPad.pad0((sat.raan * RAD2DEG).toFixed(4), 8);
    argPe = stringPad.pad0((parseFloat(argPe) / 10).toFixed(4), 8);

    // Create the new TLEs
    const TLE1 = '1 ' + sat.sccNum + 'U ' + intl + ' ' + epochyr + epochday + TLE1Ending;
    const TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + raan + ' ' + ecen + ' ' + argPe + ' ' + meana + ' ' + meanmo + '    10';

    // Calculate the orbit
    const satrec = <SatelliteRecord>satellite.twoline2satrec(TLE1, TLE2);

    // Check the orbit
    const results = getOrbitByLatLonPropagate(now, satrec, PropagationOptions.ArgumentOfPerigee);
    if (results === PropagationResults.Success) {
      sat.TLE1 = TLE1;
      sat.TLE2 = TLE2;
      newArgPer = argPe;
    }
    return results;
  };

  /**
   * Rotating the mean anomaly adjusts the latitude (and longitude) of the satellite.
   * @param {number} raan - This is the right ascension of the ascending node (where it rises above the equator relative to a specific star)
   * @param {number} raanOffsetIn - This allows the main thread to send a guess of the raan
   * @returns {PropagationResults} This number tells the main loop what to do next
   */
  const raanCalc = (raan: number, raanOffsetIn: number): PropagationResults => {
    const origRaan = raan;
    raan = raan / 100;
    raan = raan > 360 ? raan - 360 : raan;

    const raanStr = stringPad.pad0(raan.toFixed(4), 8);

    // If we adjusted argPe use the new one - otherwise use the old one
    const argPe = newArgPer ? stringPad.pad0((parseFloat(newArgPer) / 10).toFixed(4), 8) : stringPad.pad0((sat.argPe * RAD2DEG).toFixed(4), 8);

    const TLE1 = '1 ' + sat.sccNum + 'U ' + intl + ' ' + epochyr + epochday + TLE1Ending; // M' and M'' are both set to 0 to put the object in a perfect stable orbit
    const TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + raanStr + ' ' + ecen + ' ' + argPe + ' ' + newMeana + ' ' + meanmo + '    10';

    const satrec = <SatelliteRecord>satellite.twoline2satrec(TLE1, TLE2);
    const results = getOrbitByLatLonPropagate(now, satrec, PropagationOptions.RightAscensionOfAscendingNode);

    // If we have a good guess of the raan, we can use it, but need to apply the offset to the original raan
    if (results === PropagationResults.Success) {
      raan = origRaan / 100 + raanOffsetIn;
      raan = raan > 360 ? raan - 360 : raan;
      raan = raan < 0 ? raan + 360 : raan;

      const _raanStr = stringPad.pad0(raan.toFixed(4), 8);

      const _TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + _raanStr + ' ' + ecen + ' ' + argPe + ' ' + newMeana + ' ' + meanmo + '    10';

      sat.TLE1 = TLE1;
      sat.TLE2 = _TLE2;
    }
    return results;
  };

  const getOrbitByLatLonPropagate = (nowIn: Date, satrec: SatelliteRecord, type: PropagationOptions): PropagationResults => {
    const { m, gmst } = calculateTimeVariables(nowIn, satrec);
    const positionEci = <EciVec3>satellite.sgp4(satrec, m).position;
    if (isNaN(positionEci.x) || isNaN(positionEci.y) || isNaN(positionEci.z)) {
      return PropagationResults.Error;
    }
    const gpos = satellite.eciToGeodetic(positionEci, gmst);

    let { lat, lon, alt } = gpos;
    lat = satellite.degreesLat(lat);
    lon = satellite.degreesLong(lon);

    // Set it the first time
    lastLat = lastLat ? lastLat : lat;

    if (type === PropagationOptions.MeanAnomaly) {
      if (lat === lastLat) {
        return 0; // Not enough movement, skip this
      }

      if (lat > lastLat) {
        currentDirection = 'N';
      }
      if (lat < lastLat) {
        currentDirection = 'S';
      }

      lastLat = lat;
    }

    if (type === PropagationOptions.MeanAnomaly && lat > goalLat - 0.15 && lat < goalLat + 0.15) {
      return PropagationResults.Success;
    }

    if (type === PropagationOptions.RightAscensionOfAscendingNode && lon > goalLon - 0.15 && lon < goalLon + 0.15) {
      return PropagationResults.Success;
    }

    if (type === PropagationOptions.ArgumentOfPerigee && alt > goalAlt - 30 && alt < goalAlt + 30) {
      return PropagationResults.Success;
    }

    // If current latitude greater than 11 degrees off rotate meanA faster
    if (type === PropagationOptions.MeanAnomaly && !(lat > goalLat - 11 && lat < goalLat + 11)) {
      return PropagationResults.Far;
    }

    // If current longitude greater than 11 degrees off rotate raan faster
    if (type === PropagationOptions.RightAscensionOfAscendingNode && !(lon > goalLon - 11 && lon < goalLon + 11)) {
      return PropagationResults.Far;
    }

    // If current altitude greater than 100 km off rotate augPerigee faster
    if (type === PropagationOptions.ArgumentOfPerigee && (alt < goalAlt - 100 || alt > goalAlt + 100)) {
      return PropagationResults.Far;
    }

    return PropagationResults.Near;
  };
  // //////////////////////////////////////////////////////////////////////////////
  // //////////////////////////////////////////////////////////////////////////////

  let argPerCalcResults: PropagationResults;
  let meanACalcResults: PropagationResults;
  let raanCalcResults: PropagationResults;
  /** Rotate Mean Anomaly 0.1 Degree at a Time for Up To 520 Degrees */
  for (let i = 0; i < 520 * 10; i += 1) {
    meanACalcResults = meanaCalc(i);
    if (meanACalcResults === PropagationResults.Success) {
      if (currentDirection !== goalDirection) {
        // Move 2 Degrees ahead in the orbit to prevent being close on the next lattiude check
        // This happens when the goal latitude is near the poles
        i += 20; // NOSONAR
      } else {
        break; // Stop changing the Mean Anomaly
      }
    }
    if (meanACalcResults === PropagationResults.Far) {
      i += 100; // NOSONAR
    }
  }

  if (meanACalcResults !== PropagationResults.Success) {
    return ['Error', 'Failed to find a solution for Mean Anomaly'];
  }

  // ===== Argument of Perigee Loop =====
  // Applies to eccentric orbits
  if (goalAlt) {
    // Don't Bother Unless Specifically Requested
    meanACalcResults = 0; // Reset meanACalcResults
    for (let i = 0; i < 360 * 10; i += 1) {
      /** Rotate ArgPer 0.1 Degree at a Time for Up To 400 Degrees */
      argPerCalcResults = argPerCalc(i.toString());
      if (argPerCalcResults === PropagationResults.Success) {
        // DEBUG:
        // console.log('Found Correct Alt');
        if (meanACalcResults === PropagationResults.Success) {
          // DEBUG:
          // console.log('Found Correct Lat');

          // DEBUG:
          // console.log('Up Or Down: ' + upOrDown);
          if (currentDirection === goalDirection) {
            // If Object is moving in the goal direction (upOrDown)
            break; // Stop changing ArgPer
          }
        } else {
          // DEBUG:
          // console.log('Found Wrong Lat');
        }
      } else {
        // DEBUG:
        // console.log('Failed Arg of Per Calc');
      }
      if (argPerCalcResults === PropagationResults.Far) {
        // Change ArgPer faster
        i += 5 * 10; // NOSONAR
      }
      if (argPerCalcResults === PropagationResults.Error) {
        return ['Error', 'Failed to find a solution for Argument of Perigee'];
      }

      // ===== Mean Anomaly Loop =====
      for (let j = 0; j < 520 * 10; j += 1) {
        /** Rotate Mean Anomaly 0.1 Degree at a Time for Up To 400 Degrees */
        meanACalcResults = meanaCalc(j);
        if (meanACalcResults === PropagationResults.Success) {
          if (currentDirection !== goalDirection) {
            // If Object is moving opposite of the goal direction (upOrDown)
            // Move 2 Degrees ahead in the orbit to prevent being close on the next lattiude check
            j = j + 20; // NOSONAR
          } else {
            break; // Stop changing the Mean Anomaly
          }
        }
        j = meanACalcResults === PropagationResults.Far ? j + 100 : j; // NOSONAR
        if (meanACalcResults === PropagationResults.Error) {
          return ['Error', ''];
        }
      }
    }

    if (argPerCalcResults !== PropagationResults.Success) {
      return ['Error', 'Failed to find a solution for Argument of Perigee'];
    }
  }

  // ===== Right Ascension Loop =====
  for (let i = 0; i < 520 * 100; i += 1) {
    // 520 degress in 0.01 increments TODO More precise?
    raanCalcResults = raanCalc(i, raanOffset);
    if (raanCalcResults === PropagationResults.Success) {
      break;
    }
    if (raanCalcResults === PropagationResults.Far) {
      i += 10 * 100; // NOSONAR
    }
  }

  if (raanCalcResults !== PropagationResults.Success) {
    return ['Error', 'Failed to find a solution for Right Ascension of Ascending Node'];
  }

  return [sat.TLE1, sat.TLE2];
};
