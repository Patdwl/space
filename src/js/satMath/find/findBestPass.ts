import { DEG2RAD, MINUTES_PER_DAY, TAU } from '@app/js/lib/constants';
import { calcSatrec } from '@app/js/satSet/catalogSupport/calcSatrec';
import { SatelliteRecord } from 'ootk';
import { keepTrackApi } from '../../api/keepTrackApi';
import { lookanglesRow, SatObject, SensorObject } from '../../api/keepTrackTypes';
import { getRae } from '../calc/getRae';
import { checkIsInView } from '../lookangles/checkIsInView';

export const findBestPass = (sat: SatObject, sensors: SensorObject[]): lookanglesRow[] => {
  const { timeManager, uiManager } = keepTrackApi.programs;

  // Check if there is a sensor
  if (sensors.length <= 0 || !sensors[0] || typeof sensors[0].obsminaz == 'undefined') {
    uiManager.toast(`Sensor's format incorrect. Did you select a sensor first?`, 'critical');
    return [];
  }
  sensors[0].observerGd = {
    // Array to calculate look angles in propagate()
    lat: sensors[0].lat * DEG2RAD,
    lon: sensors[0].lon * DEG2RAD,
    alt: sensors[0].alt,
  };

  // TOOD: Instead of doing the first sensor this should return an array of TEARRs for all sensors.
  const sensor = sensors[0];

  const simulationTime = timeManager.simulationTimeObj;
  let offset = 0;

  var satrec = calcSatrec(sat);
  var lookanglesTable = []; // Iniially no rows to the table

  let looksInterval = 5;
  let looksLength = 7;

  // Setup flags for passes
  let score = 0;
  let sAz = <string | null>null;
  let sEl = <string | null>null;
  let srng = <string | null>null;
  let sTime = <Date | null>null;
  let passMinrng = sensor.obsmaxrange; // This is set each look to find minimum rng (start at max rng)
  let passMaxEl = 0;
  let start3 = false;
  let stop3 = false;

  let orbitalPeriod = MINUTES_PER_DAY / ((satrec.no * MINUTES_PER_DAY) / TAU); // Seconds in a day divided by mean motion

  const _propagateBestPass = (now: Date, satrecIn: SatelliteRecord): lookanglesRow => {
    let aer = getRae(now, satrecIn, sensor);
    let isInFOV = checkIsInView(sensor, aer);

    if (isInFOV) {
      // Previous Pass to Calculate first line of coverage
      const now1 = timeManager.getOffsetTimeObj(offset - looksInterval * 1000, simulationTime);
      let aer1 = getRae(now1, satrecIn, sensor);

      let isInFOV1 = checkIsInView(sensor, aer1);
      if (!isInFOV1) {
        // if it starts around 3
        if (aer.el <= 3.5) {
          start3 = true;
        }

        // First Line of Coverage
        sTime = now;
        sAz = aer.az.toFixed(0);
        sEl = aer.el.toFixed(1);
        srng = aer.rng.toFixed(0);
      } else {
        // Next Pass to Calculate Last line of coverage
        let _now1 = timeManager.getOffsetTimeObj(offset + looksInterval * 1000, simulationTime);
        aer1 = getRae(_now1, satrecIn, sensor);

        isInFOV1 = checkIsInView(sensor, aer1);
        if (!isInFOV1) {
          // if it stops around 3
          stop3 = aer.el <= 3.5;

          // Skip pass if satellite is in track right now
          if (sTime == null) {
            return {
              sortTime: null,
              scc: null,
              score: null,
              startDate: null,
              startTime: null,
              startAz: null,
              startEl: null,
              startrng: null,
              stopDate: null,
              stopTime: null,
              stopAz: null,
              stopEl: null,
              stoprng: null,
              tic: null,
              minrng: null,
              passMaxEl: null,
            };
          }

          score = Math.min((((now.getTime() - sTime.getTime()) / 1000 / 60) * 10) / 8, 10); // 8 minute pass is max score
          let elScore = Math.min((passMaxEl / 50) * 10, 10); // 50 el or above is max score

          // elScore -= Math.max((passMaxEl - 50) / 5, 0); // subtract points for being over 50 el
          elScore *= start3 && stop3 ? 2 : 1; // Double points for start and stop at 3
          score += elScore;
          score += Math.min((10 * 750) / passMinrng, 10); // 750 or less is max score

          // score -= Math.max((750 - passMinrng) / 10, 0); // subtract points for being closer than 750
          let tic = 0;
          tic = (now.getTime() - sTime.getTime()) / 1000 || 0;

          // Last Line of Coverage
          return {
            sortTime: sTime.getTime(),
            scc: satrecIn.satnum,
            score: score,
            startDate: sTime,
            startTime: sTime,
            startAz: sAz,
            startEl: sEl,
            startrng: srng,
            stopDate: now,
            stopTime: now,
            stopAz: aer.az.toFixed(0),
            stopEl: aer.el.toFixed(1),
            stoprng: aer.rng.toFixed(0),
            tic: tic,
            minrng: passMinrng.toFixed(0),
            passMaxEl: passMaxEl.toFixed(1),
          };
        }
      }
      // Do this for any pass in coverage
      if (passMaxEl < aer.el) passMaxEl = aer.el;
      if (passMinrng > aer.rng) passMinrng = aer.rng;
    }
    return {
      sortTime: null,
      scc: null,
      score: null,
      startDate: null,
      startTime: null,
      startAz: null,
      startEl: null,
      startrng: null,
      stopDate: null,
      stopTime: null,
      stopAz: null,
      stopEl: null,
      stoprng: null,
      tic: null,
      minrng: null,
      passMaxEl: null,
    };
  };

  for (let i = 0; i < looksLength * 24 * 60 * 60; i += looksInterval) {
    // lookanglesInterval in seconds
    offset = i * 1000; // Offset in seconds (msec * 1000)
    const now = timeManager.getOffsetTimeObj(offset, simulationTime);
    if (lookanglesTable.length <= 5000) {
      // Maximum of 1500 lines in the look angles table
      const _lookanglesRow = _propagateBestPass(now, satrec);
      // If data came back...
      if (_lookanglesRow.score !== null) {
        lookanglesTable.push(_lookanglesRow); // Update the table with looks for this 5 second chunk and then increase table counter by 1

        // Reset flags for next pass
        score = 0;
        sAz = null;
        sEl = null;
        srng = null;
        sTime = null;
        passMinrng = sensor.obsmaxrange; // This is set each look to find minimum rng
        passMaxEl = 0;
        start3 = false;
        stop3 = false;
        // Jump 3/4th to the next orbit
        i = i + orbitalPeriod * 60 * 0.75; // NOSONAR
      }
    }
  }

  return lookanglesTable;
};
