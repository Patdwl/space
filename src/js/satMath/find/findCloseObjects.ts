import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { EciPos, SatObject } from '@app/js/api/keepTrackTypes';
import { getUnique } from '@app/js/lib/helpers';
import { getEci } from '../calc/getEci';

let searchStrCache: string;
export const findCloseObjects = () => {
  if (searchStrCache) return searchStrCache;
  const { satSet } = keepTrackApi.programs;
  const searchRadius = 50; // km

  let csoList = [];
  let satList = <SatObject[]>[];

  // Loop through all the satellites
  for (let i = 0; i < satSet.orbitalSats; i++) {
    // Get the satellite
    const sat = satSet.getSat(i);
    // Avoid unnecessary errors
    if (typeof sat.TLE1 == 'undefined') continue;
    // Only look at satellites in LEO
    // if (sat.apogee > 5556) continue;
    // Find where the satellite is right now
    if (typeof sat.position === 'undefined') {
      sat.position = <EciPos>getEci(sat, new Date()).position || { x: 0, y: 0, z: 0 };
    }
    // If it fails, skip it
    if (isNaN(sat.position.x) || isNaN(sat.position.y) || isNaN(sat.position.z)) continue;
    if (sat.position === { x: 0, y: 0, z: 0 }) continue;
    // Add the satellite to the list
    satList.push(sat);
  }

  // Remove duplicates
  satList = getUnique(satList);

  // Sort satList by position.x property
  satList.sort((a, b) => a.position.x - b.position.x);

  // Loop through all the satellites with valid positions
  let i = 0;
  const satListLen = satList.length;
  for (i = 0; i < satListLen; i++) {
    const sat1 = satList[i];
    const pos1 = sat1.position;

    // Calculate the area around the satellite
    const posXmin = pos1.x - searchRadius;
    const posXmax = pos1.x + searchRadius;
    const posYmin = pos1.y - searchRadius;
    const posYmax = pos1.y + searchRadius;
    const posZmin = pos1.z - searchRadius;
    const posZmax = pos1.z + searchRadius;

    // Loop through the list again
    let j = 0;
    for (j = Math.max(0, i - 200); j < satListLen; j++) {
      const sat2 = satList[j]; // Get the second satellite
      if (sat1 == sat2) continue; // Skip the same satellite
      const pos2 = sat2.position; // Get the second satellite's position

      // Satellites are in order of x position so once we exceed the maxX, we can stop
      if (pos2.x > posXmax) break;
      // Check to see if the second satellite is in the search area
      if (pos2.x < posXmax && pos2.x > posXmin && pos2.y < posYmax && pos2.y > posYmin && pos2.z < posZmax && pos2.z > posZmin) {
        // Add the second satellite to the list if it is close
        csoList.push({ sat1: sat1, sat2: sat2 });
      }
    }
  }

  let csoListUnique = getUnique(csoList);

  const csoStrArr = []; // Clear CSO List

  // Loop through the possible CSOs
  for (let i = 0; i < csoListUnique.length; i++) {
    // Calculate the first CSO's position 30 minutes later
    let sat = csoListUnique[i].sat1;
    let eci = getEci(sat, new Date(Date.now() + 1000 * 60 * 30));
    if (eci.position === { x: 0, y: 0, z: 0 }) continue;
    csoListUnique[i].sat1.position = eci.position;

    // Calculate the second CSO's position 30 minutes later
    sat = csoListUnique[i].sat2;
    eci = getEci(sat, new Date(Date.now() + 1000 * 60 * 30));
    if (eci.position === { x: 0, y: 0, z: 0 }) continue;
    sat.position = eci.position;
    csoListUnique[i].sat2.position = eci.position;
  }

  // Loop through the CSOs
  for (let i = 0; i < csoListUnique.length; i++) {
    // Check the first CSO
    let sat1 = csoListUnique[i].sat1;
    let pos1 = sat1.position;
    if (typeof pos1 == 'undefined') continue;

    // Calculate the area around the CSO
    let posXmin = pos1.x - searchRadius;
    let posXmax = pos1.x + searchRadius;
    let posYmin = pos1.y - searchRadius;
    let posYmax = pos1.y + searchRadius;
    let posZmin = pos1.z - searchRadius;
    let posZmax = pos1.z + searchRadius;

    // Get the second CSO object
    let sat2 = csoListUnique[i].sat2;
    let pos2 = sat2.position;
    if (typeof pos2 == 'undefined') continue;

    // If it is still in the search area, add it to the list
    if (pos2.x < posXmax && pos2.x > posXmin && pos2.y < posYmax && pos2.y > posYmin && pos2.z < posZmax && pos2.z > posZmin) {
      csoStrArr.push(sat1.sccNum);
      csoStrArr.push(sat2.sccNum);
    }
  }

  // Generate the search string
  const csoListUniqueArr = Array.from(new Set(csoStrArr));
  let searchStr = '';
  for (let i = 0; i < csoListUniqueArr.length; i++) {
    if (i == csoListUniqueArr.length - 1) {
      searchStr += csoListUniqueArr[i];
    } else {
      searchStr += csoListUniqueArr[i] + ',';
    }
  }

  // Dont need to do this math more than once
  searchStrCache = searchStr;
  return searchStr; // csoListUnique;
};
