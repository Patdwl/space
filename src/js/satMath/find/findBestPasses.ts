import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { SensorObject } from '@app/js/api/keepTrackTypes';
import { saveCsv } from '@app/js/lib/helpers';
import { findBestPass } from './findBestPass';

export const findBestPasses = (sats: string, sensor: SensorObject) => {
  const { satSet } = keepTrackApi.programs;

  sats = sats.replace(/ /gu, ',');
  const satArray = sats.split(',');
  let tableSatTimes = [];
  for (let i = 0; i < satArray.length; i++) {
    try {
      const satId = satArray[i];
      if (typeof satId == 'undefined' || satId == null || satId === '' || satId === ' ') continue;
      const sat = satSet.getSatFromObjNum(parseInt(satId));
      const satPasses = findBestPass(sat, [sensor]);
      for (let s = 0; s < satPasses.length; s++) {
        tableSatTimes.push(satPasses[s]);
        // }
      }
    } catch (e) {
      console.debug(e);
    }
  }
  tableSatTimes.sort((a, b) => b.sortTime - a.sortTime);
  tableSatTimes.reverse();
  tableSatTimes.forEach((v) => {
    delete v.sortTime;
  });

  for (let i = 0; i < tableSatTimes.length; i++) {
    tableSatTimes[i].startDate = tableSatTimes[i].startDate.toISOString().split('T')[0];
    tableSatTimes[i].startTime = tableSatTimes[i].startTime.toISOString().split('T')[1].split('.')[0];
    tableSatTimes[i].stopDate = tableSatTimes[i].stopDate.toISOString().split('T')[0];
    tableSatTimes[i].stopTime = tableSatTimes[i].stopTime.toISOString().split('T')[1].split('.')[0];
  }

  saveCsv(tableSatTimes, 'bestSatTimes');
};
