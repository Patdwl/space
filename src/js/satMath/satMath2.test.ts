// TODO: Jest snapshots are currently dependent on local computer time. They
// should be updated to use the same time on all computers.

import { defaultSat, defaultSensor, keepTrackApiStubs } from '../api/apiMocks';
import { keepTrackApi } from '../api/keepTrackApi';
import { KeepTrackPrograms } from '../api/keepTrackTypes';
import * as calculateLookAngles from './calc/calculateLookAngles';
import * as getEci from './calc/getEci';
import { getOrbitByLatLon } from './calc/getOrbitByLatLon';
import * as findBestPass from './find/findBestPass';
import { findBestPasses } from './find/findBestPasses';
import { findCloseObjects } from './find/findCloseObjects';
import { findNearbyObjectsByOrbit } from './find/findNearbyObjectsByOrbit';
import { eci2ll, eci2rae } from './transforms';
import * as lookAngles2Ecf from './transforms/lookAngles2Ecf';

keepTrackApi.programs = <KeepTrackPrograms>(<unknown>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs });
const dateNow = new Date(2022, 0, 1);
dateNow.setUTCHours(0, 0, 0, 0);

// @ponicode
describe.skip('satMath.findCloseObjects', () => {
  test('0', () => {
    let result: any = findCloseObjects();
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('satMath.getOrbitByLatLon', () => {
  test('0', () => {
    let result: any = getOrbitByLatLon(defaultSat, 0, 0, 'N', dateNow, 1000, 0);
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('satMath.calculateLookAngles', () => {
  test('0', () => {
    calculateLookAngles.calculateLookAngles(defaultSat, [defaultSensor]);
  });
});

// @ponicode
describe('satMath.findBestPasses', () => {
  test('0', () => {
    let result: any = findBestPasses('25544', defaultSensor);
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('satMath.findBestPass', () => {
  test('0', () => {
    findBestPass.findBestPass(defaultSat, [defaultSensor]);
  });
});

// @ponicode
describe('satMath.eci2Rae', () => {
  test('0', () => {
    eci2rae(dateNow, [10000, 10000, 10000], defaultSensor);
  });
});

// @ponicode
describe('satMath.getEci', () => {
  test('0', () => {
    getEci.getEci(defaultSat, dateNow);
  });
});

// @ponicode
describe.skip('satMath.findNearbyObjectsByOrbit', () => {
  test('0', () => {
    let result: any = findNearbyObjectsByOrbit(defaultSat);
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('satMath.lookAngles2Ecf', () => {
  test('0', () => {
    let result: any = lookAngles2Ecf.lookAngles2ecf(47, 20, 2000, 41, -71, 0);
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let result: any = lookAngles2Ecf.lookAngles2ecf(167, 20, 5000, 41, -71, 0);
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('satMath.eci2ll', () => {
  test('0', () => {
    eci2ll(1000, 2000, 4000);
  });

  test('1', () => {
    eci2ll(-1000, 5000, 0);
  });
});
