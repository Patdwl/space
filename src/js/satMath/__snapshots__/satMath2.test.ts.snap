// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`satMath.findBestPasses 0 1`] = `undefined`;

exports[`satMath.getOrbitByLatLon 0 1`] = `
Array [
  "Error",
  "Failed to find a solution for Mean Anomaly",
]
`;

exports[`satMath.lookAngles2Ecf 0 1`] = `
Object {
  "x": 3399.2987560938654,
  "y": 6013.8733407148075,
  "z": -2379.043164190548,
}
`;

exports[`satMath.lookAngles2Ecf 1 1`] = `
Object {
  "x": 3697.567569850421,
  "y": 7959.494204847528,
  "z": 3243.7908022872425,
}
`;
